llvm/lib/Target/X86/labs/zakharov_artem/X86ZakharovInstCounterPass.cpp

```diff
@@ -0,0 +1,43 @@
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+
+using namespace llvm;
+
+namespace {
+
+class X86ZakharovInstCounterPass : public MachineFunctionPass {
+public:
+  static char ID;
+  X86ZakharovInstCounterPass() : MachineFunctionPass(ID) {}
+  bool runOnMachineFunction(llvm::MachineFunction &MF) override {
+    auto *GV = MF.getFunction().getParent()->getNamedGlobal("ic");
+    if (!GV) {
+      return false;
+    }
+
+    auto DL = MF.front().begin()->getDebugLoc();
+    auto *TII = MF.getSubtarget().getInstrInfo();
+
+    for (auto &MBB : MF) {
+      int Counter = std::distance(MBB.begin(), MBB.end());
+      auto Place = MBB.getFirstTerminator();
+      if (Place != MBB.begin()) {
```

Comment:
Why do you go step backward here?

Reply:
If the terminator is the JCC, then the instruction to increase the counter will be added between JCC and CMP. I thought that the ADD instruction might change the flags set by the CMP and decided to add it above

Reply:
Ok, then it would be better to explicitly check terminator instruction type and decrease this iterator only in case of jcc. And a comment describing this logic would be very valuable for future readers of that code.